### PR TITLE
SQL AAD Setup

### LINKED ISSUE #55: Create & Configure Feature data store
Description

As a data engineer I want a central place where I could write features into, so other FlowEHR personas (e.g. data scientists, software engineers) could read from the Gold data store. 

Acceptance Criteria

- SQL Server Database is created
- Write permissions from a data pipeline into a Gold Data Store (SQL Server Database) are configured

### LINKED ISSUE #95: SQL Feature Data Store: AAD Setup
- [ ] Create a user managed identity for the SQL store
- [ ] Grant the identity needed permissions in the tenancy (or add to a group that has those permissions)
- [ ] Add the SPN as a user in the database, via python script, with a new schema it owns
- [ ] Show how a databricks notebook could use the SPN secrets in the secret scope to AAD-connect to the SQL store and query / create some data
- [ ] Document any manual steps

Some relevant docs here: 
- https://learn.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-service-principal-tutorial
- https://learn.microsoft.com/en-us/sql/connect/python/pyodbc/step-3-proof-of-concept-connecting-to-sql-using-pyodbc?view=sql-server-ver16

### PR DESCRIPTION
**Pull request by damoodamoo**
# Resolves #55 and Resolves #95

- Adds `make ci-auth`, using local terraform state, intended to be run by a Admin prior to setting up CI. 
- Creates a SQL Owner AAD App, and assigns the associated service principal as the AAD owner of the SQL database. 
- Creates a System Managed Identity for the SQL Server, and assigns `user.read.all | groupmember.read.all | application.read.all` as least-privilege as per the docs
- Adds a script which uses the SQL Owner AAD app credentials + pyodbc with SQL Driver 18 to connect to the new database and create a `LOGIN` and `USER` for the databricks-sql aad app, and assigns that user `db_owner` on the feature database.
- Adds a sample/demo notebook to show a pyodbc query from databricks to SQL, used for validation - which pulls the SQL + AAD details from databricks secrets, and uses them to query a table.

- [X] Test from local with high level AAD perms
- [x] Test from damoodamoo ci fork in private subscription
- [x] Assign permissions in UCLH tenancy + test in UCLH dev
- [x] Document permissions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,25 +65,31 @@ For the full reference of possible configuration values, see the [config schema
 CI deployment workflows are run in [Github environments](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment). These should
 be created in a private repository created from this template repository.
 
-1. Create a service principal
+This step will create an AAP Application and Service Principal in the specified tenancy, and grant that service principal permissions in Azure and AAD needed for deployment. These are detailed below.
```
Comment (jjgriff93) [inline README.md:68]:
```suggestion
This step will create an AAD Application and Service Principal in the specified tenancy, and grant that service principal permissions in Azure and AAD needed for deployment. These are detailed below. 
```

File: README.md [hunk 2]
```diff
@@ -65,25 +65,31 @@ For the full reference of possible configuration values, see the [config schema
...
+
+    - Copy the block of JSON from the terminal for the next step.
 
 
 2. Create and populate a GitHub environment
 
     Add an environment called `Infra-Test` with the following secrets:
 
-    - `AZURE_CREDENTIALS`: json containing the credentials of the service principal in the format:
+    - `AZURE_CREDENTIALS`: json containing the credentials of the service principal in the format. This should have been output for you in the correct format in the previous step:
```
Comment (jjgriff93) [inline README.md:92]:
```suggestion
    - `AZURE_CREDENTIALS`: json containing the credentials of the service principal (outputted from the previous step) in the following format:
```

File: infrastructure/ci-auth/main.tf [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+# Get the MS Graph app 
+resource "azuread_service_principal" "msgraph" {
+  application_id = data.azuread_application_published_app_ids.well_known.result.MicrosoftGraph
+  use_existing   = true
+}
+
+resource "azuread_application" "ci_app" {
+  display_name = "sp-flowehr-cicd-${lower(var.naming_suffix)}"
+  owners       = [data.azurerm_client_config.current.object_id]
+  required_resource_access {
```
Comment (jjgriff93) [inline infrastructure/ci-auth/main.tf:24]:
```suggestion
  
  required_resource_access {
```

File: infrastructure/ci-auth/main.tf [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    resource_app_id = data.azuread_application_published_app_ids.well_known.result.MicrosoftGraph
+
+    resource_access {
+      id   = azuread_service_principal.msgraph.app_role_ids["Application.ReadWrite.All"]
+      type = "Role"
+    }
+    resource_access {
```
Comment (jjgriff93) [inline infrastructure/ci-auth/main.tf:31]:
```suggestion
    
    resource_access {
```

**Review by jjgriff93 — APPROVED** [review]
Great work!

File: infrastructure/transform/feature-data-store.tf [hunk 5]
```diff
@@ -47,7 +88,7 @@ resource "azurerm_mssql_server_transparent_data_encryption" "sql_server_features
 }
 
 # Azure SQL database, basic + small for dev
-# TODO: Rightsize for prod -> https://github.com/UCLH-Foundry/FlowEHR/issues/63
+# TODO: Rightsize for prod -> https://github.com/UCLH-Foundry/issues/63
```
Comment (t-young31) [inline infrastructure/transform/feature-data-store.tf:91]: ?

File: infrastructure/transform/feature-data-store.tf [hunk 6]
```diff
@@ -60,21 +101,63 @@ resource "azurerm_mssql_database" "feature_database" {
...
+    command = <<EOF
+      python ../../scripts/sql/create_sql_user.py $sql_server $database $app_id $app_secret $app_name
+      python ../../scripts/sql/load_csv_data.py $sql_server $database $app_id $app_secret "../../scripts/sql/nhsd-diabetes.csv" "diabetes"
+    EOF
+    environment = {
+      sql_server = azurerm_mssql_server.sql_server_features.fully_qualified_domain_name
+      database   = azurerm_mssql_database.feature_database.name
+      app_id     = azuread_application.flowehr_sql_owner.application_id
+      app_secret = azuread_application_password.flowehr_sql_owner.value
+      app_name   = local.databricks_app_name
```
Comment (t-young31) [inline infrastructure/transform/feature-data-store.tf:131]: why both as env vars and cli args?
fwiw I'd vote env vars for consistency with the other python scripts

File: scripts/sql/CSV-README.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# Note on CSV Dataset committed in this directory
```
Comment (t-young31) [inline scripts/sql/CSV-README.md:1]: maybe just call this README.md for some visibility?

File: infrastructure/transform/feature-data-store.tf (hunk 5 above)
Comment (damoodamoo) [reply infrastructure/transform/feature-data-store.tf:91]: weird!

File: scripts/sql/create_sql_user.py [hunk 8]
```diff
@@ -13,13 +13,12 @@
...
 
 import pyodbc
-import sys
 
-server = sys.argv[1]
-database = sys.argv[2]
-client_id = sys.argv[3]
-client_secret = sys.argv[4]
-login_to_create = sys.argv[5]
+server = "${SERVER}"
```
Comment (t-young31) [inline scripts/sql/create_sql_user.py]: does this work in .py???

Comment (t-young31) [reply scripts/sql/create_sql_user.py]: hehe. snap

File: infrastructure/transform/feature-data-store.tf (hunk 6 above)
Comment (damoodamoo) [reply infrastructure/transform/feature-data-store.tf]: yea ok, switched to env vars entirely, i guess you could still pass them in the cmd line if you want to call the script locally

**Comment by damoodamoo** [comment]
/test-force-approve

PR Env for CI got into bad state - ran in separate PR here: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4286381496/jobs/7465880646

**Pull request merged**